stan/math/prim/functor/hcubature.hpp

```diff
@@ -0,0 +1,521 @@
+    for (size_t j = 0; j != k; j++)
+      temp[c[j] - 1] = lambda;
+    p.push_back(temp);
+  }
+  free(c);
+}
+
+template <typename T_index, typename T_k, typename T_lambda, typename T_n,
+          typename T_c, typename T_temp>
+void increment(T_index& index, const T_k& k, const T_lambda& lambda,
+               const T_n& n, const T_c& c, T_temp& temp) {
+  // temp size n, all elements initially zero
+  if (index.size() == 0) {
+    index.push_back(false);
+    for (size_t j = 0; j != k; j++)
+      temp[c[j] - 1] = lambda;
+    return;
+  }
+  T_k first_zero = 0;
+  while ((first_zero < index.size()) && index[first_zero])
+    first_zero++;
+  if (first_zero == index.size()) {
+    index.flip();
+    for (size_t j = 0; j != index.size(); j++)
+      temp[c[j] - 1] *= -1;
+    index.push_back(true);
+    temp[c[index.size() - 1] - 1] = -lambda;
+  } else {
+    for (size_t i = 0; i != first_zero + 1; i++) {
+      //           index[i] = !index[i];
+      if (index[i])
+        index[i] = 0;
+      else
+        index[i] = 1;
+      temp[c[i] - 1] *= -1;
+    }
+  }
+}
+
+template <typename T_k, typename T_lambda, typename T_n, typename T_p>
+void signcombos(const T_k& k, const T_lambda& lambda, const T_n& n, T_p& p) {
+  T_k* c = reinterpret_cast<T_k*>(malloc(k * sizeof(T_k)));
+  for (size_t i = 1; i != choose(n, k) + 1; i++) {
+    T_lambda* temp = reinterpret_cast<T_lambda*>(calloc(n, sizeof(T_lambda)));
+    combination(c, n, k, i);
+    std::vector<bool> index;
+    index.clear();
+    for (size_t j = 0; j != std::pow(2, k); j++) {
+      increment(index, k, lambda, n, c, temp);
+      T_lambda* next
+          = reinterpret_cast<T_lambda*>(malloc(n * sizeof(T_lambda)));
+      memcpy(next, temp, n * sizeof(T_lambda));
+      p.push_back(next);
+    }
+    free(temp);
+  }
+  free(c);
+}
+
+template <typename F, typename T_a, typename T_b, typename T_oned,
+          typename T_pars>
+void gauss_kronrod(const F& integrand, const T_a& a, const T_b& b, T_oned& out,
+                   T_pars& pars) {
+  using T_return_type = return_type_t<T_a, T_b, T_oned, T_pars>;
+  std::vector<T_return_type> c(1, 0);
+  std::vector<T_return_type> cp(1, 0);
+  std::vector<T_return_type> cm(1, 0);
+  c[0] = 0.5 * (a + b);
+  T_return_type delta = 0.5 * (b - a);
+  T_return_type f0 = integrand(c, pars);
+  T_return_type I = f0 * wd7[7];
+  T_return_type Idash = f0 * gwd7[3];
+  for (size_t i = 0; i != 7; i++) {
+    T_return_type deltax = delta * xd7[i];
+    cp[0] = c[0] + deltax;
+    cm[0] = c[0] - deltax;
+    T_return_type fx = integrand(cp, pars);
+    T_return_type temp = integrand(cm, pars);
+    fx += temp;
+    I += fx * wd7[i];
+    if (i % 2 == 1)
+      Idash += fx * gwd7[i / 2];
+  }
+  T_return_type V = fabs(delta);
+  I *= V;
+  Idash *= V;
+  out.result = I;
+  out.err = fabs(I - Idash);
+  std::vector<T_return_type>().swap(c);
+  std::vector<T_return_type>().swap(cp);
+  std::vector<T_return_type>().swap(cm);
```

Comment:
So you suggest to delete all 11 swaps in the document and don't insert a clear or shrink_to_fit command?

Reply:
When I said the vectors are destroyed soon anyway I was thinking of these three specifically but looking at it again, yes, all 11 swaps are just freeing allocations that would free anyway. The only one that doesn't occur at the end of a scope is the one on line 312
```cpp
  std::vector<T_return_type>().swap(cc);
```
Removing this will postpone the freeing until the end of the function but I don't see anything that particularly cares exactly when it's freed.

So yes, I'd say completely remove them all.

Reply:
Comes in next commit.